pkg/supportbundle/supportbundle.go

```diff
@@ -0,0 +1,310 @@
+package supportbundle
+
+import (
+	"archive/zip"
+	"bytes"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"path/filepath"
+	"runtime"
+	"runtime/pprof"
+	"sync"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/grafana/agent/pkg/build"
+	"github.com/grafana/agent/pkg/config"
+	"github.com/grafana/agent/pkg/traces"
+	zaplogfmt "github.com/jsternberg/zap-logfmt"
+	"github.com/mackerelio/go-osstat/uptime"
+	"github.com/sirupsen/logrus"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+	"gopkg.in/yaml.v3"
+)
+
+// Bundle collects all the data that is exposed as a support bundle.
+type Bundle struct {
+	meta                  []byte
+	config                []byte
+	agentMetrics          []byte
+	agentMetricsInstances []byte
+	agentMetricsTargets   []byte
+	agentLogsInstances    []byte
+	heapBuf               *bytes.Buffer
+	goroutineBuf          *bytes.Buffer
+	blockBuf              *bytes.Buffer
+	mutexBuf              *bytes.Buffer
+	cpuBuf                *bytes.Buffer
+}
+
+// Metadata contains general runtime information about the current Agent.
+type Metadata struct {
+	BuildVersion    string   `yaml:"build_version"`
+	OS              string   `yaml:"os"`
+	Architecture    string   `yaml:"architecture"`
+	Uptime          float64  `yaml:"uptime"`
+	EnabledFeatures []string `yaml:"enabled_features"`
+}
+
+// Export gathers the information required for the support bundle.
+func Export(enabledFeatures []string, cfg config.Config, srvAddress string, duration float64) (*Bundle, error) {
+	// Gather runtime metadata.
+	ut, err := uptime.Get()
+	if err != nil {
+		return nil, err
+	}
+	m := Metadata{
+		BuildVersion:    build.Version,
+		OS:              runtime.GOOS,
+		Architecture:    runtime.GOARCH,
+		Uptime:          ut.Seconds(),
+		EnabledFeatures: enabledFeatures,
+	}
+	meta, err := yaml.Marshal(m)
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal support bundle metadata: %s", err)
+	}
+	// Gather current configuration.
+	config, err := yaml.Marshal(cfg)
```

Comment:
Is this already removing the secrets from the config?

Reply:
Supposedly, yes, but I'll have to double check before enabling this endpoint.

Reply:
We went ahead with the approach of not exposing the configuration, unless the -config.enable-read-api flag is passed, so as to not provide a way to sidestep it.

We should revisit them both soon-ish, as it will make support bundles more useful.